### PR TITLE
Add funds

### PR DESCRIPTION
**Pull request by mknudsen01**
Add overall funds and then earmark some for particular groups

### CODE REVIEW

File: app/javascripts/components.jsx [hunk 1]
```diff
@@ -71,6 +138,16 @@ Group = React.createClass({
     })
   },
 
+  getAvailableFunds: function() {
+    var availableFunds = this.props.earmark;
+    if (this.props.transactions.length != 0) {
```
Comment (raorao) [inline app/javascripts/components.jsx]: is this check necessary? transactions is always an array, right?

File: app/javascripts/components.jsx [hunk 2]
```diff
@@ -71,6 +138,16 @@ Group = React.createClass({
...
+      this.props.transactions.forEach(function(transaction) {
+        availableFunds += transaction.amount;
+      })
+    }
+    return availableFunds;
```
Comment (raorao) [inline app/javascripts/components.jsx]: we should avoid using mutating temps when possible, like ```availableFunds```. instead, use a collection closure method!

```
availableFunds + this.props.transactions.reduce(function(sum,transaction) { return sum + transaction.amount })
```

File: app/javascripts/components.jsx [hunk 3]
```diff
@@ -78,13 +155,15 @@ Group = React.createClass({
 
     var maybeCreateTransaction   = this.state.isCreating ? <CreateTransaction groupName={this.props.name}/> : void 0
     var addTransactionButtonText = this.state.isCreating ? "Cancel" : "Add transaction"
-
+    var availableFunds = this.getAvailableFunds();
```
Comment (raorao) [inline app/javascripts/components.jsx]: stylistic choice, but I'd prefer if we got rid of the excess variable and moved the query function into the jsx. no reason to have excess variables.

File: app/javascripts/store.js [hunk 4]
```diff
@@ -2,11 +2,14 @@ Store = (function() {
...
     var contents = {
         groups: [],
-        transactions: []
+        transactions: [],
+        earmarks: {}
     }
 
     var transactionId = 0;
 
+    unassignedFunds = 0;
```
Comment (raorao) [inline app/javascripts/store.js]: we should probably through this into contents.

File: app/javascripts/store.js [hunk 5]
```diff
@@ -15,50 +18,64 @@ Store = (function() {
...
+
+        getUnassignedFunds: function() {
+            return unassignedFunds;
         }
     }
+
+
+    return {
+        createGroup: function(groupName) {
+            contents.groups.push({name: groupName});
```
Comment (raorao) [inline app/javascripts/store.js]: a part of me dislikes that we're double storing groups in the groups array, and in earmarks. whenever we delete a group, or change its name, or create one, we'll have to update both values. That's probably okay? I dont know.

File: app/javascripts/actions.js [hunk 6]
```diff
@@ -7,5 +7,15 @@ AppAction = {
...
   createTransaction: function(transactionData) {
     Dispatcher.dispatch( { actionType: AppAction.CREATE_TRANSACTION, transactionData: transactionData } );
+  },
+
+  ADD_FUNDS: 'ADD_FUNDS',
+  addFunds: function(amount) {
+	Dispatcher.dispatch( { actionType: AppAction.ADD_FUNDS, amount: amount } );
+  },
+
+  ADD_GROUP_FUNDS: 'ADD_GROUP_FUNDS',
```
Comment (raorao) [inline app/javascripts/actions.js]: we could unpack fundData right here for clarity. 

```
Dispatcher.dispatch( { actionType: AppAction.ADD_GROUP_FUNDS, group: fundData.group, amount: fundData.amount } );
```

arguably makes the store a little cleaner.

File: app/javascripts/app.js [hunk 7]
```diff
@@ -6,6 +6,12 @@ Dispatcher.register(function(payload) {
     case AppAction.CREATE_TRANSACTION:
       Store.createTransaction(payload.transactionData);
       break;
```
Comment (raorao) [inline app/javascripts/app.js]: debating whether its worth having these as two actions. What if a ```nil``` group value meant that the funds were added to unassigned finds? it would simplify the dispatcher's API, at the expense of adding complexity to the store. 

the nice thing about flux is that the store is very, very testable. which means you can write pretty complex functions since you can test edge cases with abandon. 

you can go too far with that, obviously. There's not right answer in this case IMO.

File: app/javascripts/store.js [hunk 8]
```diff
@@ -15,50 +18,64 @@ Store = (function() {
...
+            contents.earmarks[groupName] = 0;
+            emitChange();
+        },
+        createTransaction: function(transaction) {
+            transactionId++;
+            transaction['id'] = transactionId;
```
Comment (raorao) [inline app/javascripts/store.js]:
```
 transaction['id'] = ++transactionId;
```
solves this problem as well. 

In general, I dislike separating coupled logic into separate lines. line 62 and 63 need to happen in that order to work, and would fail in crazy ways if you messed it up. inlining would surface intent, IMO.

File: app/javascripts/store.js [hunk 9]
```diff
@@ -15,50 +18,64 @@ Store = (function() {
...
-            var emptyGroups = this.getGroups().map(function(group) {
+            var emptyGroups = contents.groups.map(function(group) {
                 return {
                     name: group.name,
-                    transactions: []
+                    transactions: [],
+                    earmark: contents.earmarks[group.name]
                 }
             })
-            return this.getTransactions().reduce(function(groupsArray,transaction) {
```
Comment (raorao) [inline app/javascripts/store.js]: these can be returned to using helpers, if you want. I can't tell which would be better.

**Comment by raorao** [comment]
a couple of comments. back to you.

File: app/javascripts/app.js (hunk 7 above)
Comment (mknudsen01) [inline app/javascripts/app.js]: Trying to think of why this wouldn't work. All I could come up with is what if they add a transaction in a group that never got named. But. That shouldn't happen because there should be validation that a group name is required. So, I'm thinking I'll try to implement this change. Because there really isn't any difference except that one has a group and one doesn't.

**Comment by mknudsen01** [comment]
@raorao paired, merging.
